nattraversal.go

```diff
@@ -0,0 +1,158 @@
+package main
+
+import (
+	"context"
+	"net"
+
+	"github.com/NebulousLabs/go-upnp"
+	"github.com/go-errors/errors"
+	"github.com/jackpal/gateway"
+	"github.com/jackpal/go-nat-pmp"
+)
+
+var (
+	// Compile time checks to ensure that the different NAT implementations
+	// adhere to the interface.
+	_ NATTraversal = (*upnpTraversal)(nil)
+	_ NATTraversal = (*natPMPTraversal)(nil)
+
+	// ErrMultipleNAT is an error returned when multiple NATs have been
+	// detected.
+	ErrMultipleNAT = errors.New("could not retrieve external IP: " +
+		"multiple NATs are not supported")
+)
+
+// NATTraversal is an interface that represents the different NAT traversal
+// techniques.
+type NATTraversal interface {
```

Comment:
I dig this interface and the implementations! I think we can justify moving this into its own `nat` package, since `main` is already pretty big, and most of this logic seems pretty isolated. I'm also sure other go projects would find the ability to import and use `nat.Traversal` useful :)

Reply:
Fixed.